src/browser/fetch.rs

```diff
@@ -0,0 +1,128 @@
+//! Fetch API.
+//!
+//! Our version of the Fetch API is based mostly on regular web one.
+//!
+//! There is one entry point: [`fetch`][fetch] function.
+//! It can accept both, String urls as well as [`Request`][request].
+//!
+//! To get a [`Response`][response] you need to `.await` fetch:
+//! ```rust
+//! let response = fetch("/foo").await?;
+//! ```
+//!
+//! Then you can check [`Status`][status] and extract body in various formats:
+//! ```rust
+//! let response = fetch("/foo").await?;
+//! assert!(response.status().is_ok());
+//! let body: FooStruct = response.json().await?;
+//! ```
+//!
+//! As Rust doesn't have optional arguments we have no `fetch(url, init)` version,
+//! instead you should use something like this:
+//! ```rust
+//! fetch(Request::new(url).method(Method::Post)).await
+//! ```
+//!
+//!
+//! [fetch]: ./fn.fetch.html
+//! [request]: ./struct.Request.html
+//! [response]: ./struct.Response.html
+//! [status]: ./struct.Status.html
+
+use crate::browser::Url;
+use crate::util::window;
+use serde_json;
+use std::borrow::Cow;
+use wasm_bindgen_futures::JsFuture;
+use web_sys;
+
+pub mod header;
+mod method;
+mod request;
+mod response;
+mod status;
+
+pub use header::{Header, Headers};
+pub use method::*;
+pub use request::*;
+pub use response::*;
+pub use status::*;
+
+/// Convenient type alias.
+pub type Result<T> = std::result::Result<T, FetchError>;
+
+/// The fetch functions is a main entry point of the Fetch API.
+///
+/// It start the process of fetching a resource from the network,
+/// returning a future which is fulfilled once the response is
+/// available. The future resolves to the Response object representing
+/// the response to your request. The promise does not reject on HTTP
+/// errors — it only rejects on network errors. You must use then
+/// handlers to check for HTTP errors.
+pub async fn fetch<'a>(resourse: impl Into<Resource<'a>>) -> Result<Response> {
```

Comment:
Why we need `Resource`? Isn't `impl Into<Request<'a>` enough?
We can implement 
```rust
impl <'a, T> From<T> for Request<'a>
    where
        T: Into<Cow<'a, str>>
```
and it should cover all cases (?)

Reply:
resour **c** e

Reply:
To utilize `window().fetch_with_str(&string)` and `fetch_with_request(&request.into())` so we don't have to construct request.

Reply:
I don't know if more complex public API and more Rust code (~ more potential bugs and a little bit bigger file size) is a good trade-off for this performance (?) optimization.